File: scala/131.scala
import scala.collection.mutable.ListBuffer
object Solution {
    def test(s: String): Boolean = {
        s == s.reverse
    }
    def partition(s: String): List[List[String]] = {
        if (s.isEmpty()) List(Nil)
        else {
            val ans = ListBuffer[List[String]]()
            for (i <- 1 to s.length()) {
                val sub = s.slice(0, i)
                if (test(sub)) {
                    ans ++= partition(s.slice(i, s.length())).map(res => sub :: res).toList
                }
            }
            ans.toList
        }
    }
}

File: leetcodeJava/131.java
package leetcodeJava;

class Solution {
    public List<List<String>> partition(String s) {
        if (s.length() == 0) return List.of(new LinkedList<String>());
        var res = new LinkedList<List<String>>();
        for (var i = 1; i <= s.length(); i++) {
            if (test(s.substring(0, i))) {
                var partitions = partition(s.substring(i, s.length()));
                for (var p : partitions) {
                    p.add(0, s.substring(0, i));
                    res.add(p);
                }
            }
        }
        return res;
    }
    
    private boolean test(String s) {
        for (var i = 0; i <= s.length()/2; i++) 
            if (s.charAt(i) != s.charAt(s.length()-1-i)) return false;
        return true;
    }
}
